File: algebra.hs
{-
{Algebra gameplay}

The proof gameplay consists of subsequent stages of proving group axioms
for the function by fiddling the axiom equation to show equality.
-}
module Algebra where

isTautology :: Rule -> Bool
isTautology (Rule (a,b)) = a == b

isBinding :: Expr -> Rule -> Bool
isBinding e (Rule (a,b)) = e == a || e == b

opf :: Op -> Expr -> Expr -> Expr
opf op a b = Expr (op, a, b)

{-
Group axioms main stage:

    Stability: a o b exists in G for all a, b in G
    Magma!

    Associativity: a o (b o c) = (a o b) o c
    Semigroup!

    Neutral element: a o 0 = 0 o a = a
    Monoid!

    Inverse element: a o a_inv = a_inv o a = 0
    Group! Enter Abelian bonus stage!
-}

associativity :: Op -> Rule
associativity op = (A `o` (B `o` C)) `eq` ((A `o` B) `o` C)
                  where o = opf op

rightNeutral :: Op -> Rule
rightNeutral op = (A `o` Neutral op) `eq` A
              where o = opf op

leftNeutral :: Op -> Rule
leftNeutral op = (Neutral op `o` A) `eq` A
              where o = opf op

rightInverse :: Op -> Rule
rightInverse op = (A `o` Inv (op, A)) `eq` Neutral op
              where o = opf op

leftInverse :: Op -> Rule
leftInverse op = (Inv (op, A) `o` A) `eq` Neutral op
              where o = opf op

{-
Abelian bonus stage:

    Commutativity: a o b = b o a
    Abelian group! Enter ring bonus stage!
-}

commutativity :: Op -> Rule
commutativity op = (A `o` B) `eq` (B `o` A)
                  where o = opf op

magma op = [] -- FIXME: figure out some way to prove that op :: G -> G -> G
semiGroup op = magma op ++ [associativity op]
monoid op = semiGroup op ++ [rightNeutral op, leftNeutral op]
group op = monoid op ++ [rightInverse op, leftInverse op]
abelianGroup op = group op ++ [commutativity op]

{-
Ring bonus stage:

    Show bonus function to be a semigroup!

    Distributivity of x over o:
    a x (b o c) = (a x b) o (a x c)
    (a o b) x c = (a x c) o (b x c)
    Pseudo-ring!
-}

leftDistributivity :: Op -> Op -> Rule
leftDistributivity opO opX = (A `x` (B `o` C)) `eq` ((A `x` B) `o` (B `x` C))
                             where o = opf opO
                                   x = opf opX

rightDistributivity :: Op -> Op -> Rule
rightDistributivity opO opX = (A `x` (B `o` C)) `eq` ((A `x` B) `o` (B `x` C))
                               where o = opf opO
                                     x = opf opX
{-
    Neutral element for x: a x 1 = 1 x a = a
    Ring!

    Commutativity for x: a x b = b x a
    Commutative ring! Enter field bonus stage!

Field bonus stage:

    Inverse element for x in G \ {0}: a x a_inv = a_inv x a = 1
    Field! Superior! Shower of jewels!
-}

pseudoRing o x = abelianGroup o ++ semiGroup x ++
                 [leftDistributivity o x, rightDistributivity o x]
ring o x = pseudoRing o x ++ [rightNeutral x, leftNeutral x]
commutativeRing o x = ring o x ++ [commutativity x]
field o x = commutativeRing o x ++ [rightInverse x, leftInverse x]
{-
Whenever you show something, the equality is added to your proof inventory.
You can use items in your proof inventory to do substitutions and other manipulations.
You begin the game armed (in the tutorial) with +. Build it up from there.
-}

type ProofInventory = [Rule]

findMatchingEqualitiesAt :: Int -> Expr -> ProofInventory -> ProofInventory
findMatchingEqualitiesAt idx expr inventory =
    maybe [] (\e -> filter (matchEquality e) inventory) (subExprAt idx expr)

matchEquality :: Expr -> Rule -> Bool
matchEquality e (Rule (a,b)) = matchPattern a e || matchPattern b e

{-
{Functions}

Functions are composed of explicitly parenthesized binary operators and variables.
The names of the variables are always A and B.

E.g. the function
    f a b = a + b + 1
would be represented as
    (+ (+ a b) 1)
    Expr (Plus, (Expr (Plus, A, B), 1)

-}

type Op = String
data Expr = Expr (Op, Expr, Expr) | A | B | C | Inv (Op, Expr) | Neutral Op | Literal Int

instance Show Expr where
    show (Expr (o,l,r)) = "("++show l++" "++o++" "++show r++")"
    show A = "a"
    show B = "b"
    show C = "c"
    show (Inv (o,e)) = "inv("++o++","++show e++")"
    show (Neutral o) = "e("++o++")"
    show (Literal i) = show i

instance Eq Expr where
    Expr a == Expr b = a == b
    A == A = True
    B == B = True
    C == C = True
    Inv a == Inv b = a == b
    Neutral aop == Neutral bop = aop == bop
    Literal a == Literal b = a == b
    _ == _ = False

{-
{Rules}

The core of the gameplay is based on rewriting equations.
To rewrite an equation, you apply a rule to an expression.

E.g. the rule
    a o b = a + b + 1
would be represented as
    rule = Rule ((Expr ("o", A, B)), (Expr ("+", Expr ("+", A, B), Literal 1)))
-}

type Pattern = Expr
type Substitution = Expr

data Rule = Rule (Pattern, Substitution)
instance Show Rule where
    show (Rule (a, b)) = show a ++ " = " ++ show b

instance Eq Rule where
    Rule a == Rule b = a == b

ruleLength :: Rule -> Int
ruleLength (Rule (a, b)) = exprLength a + 1 + exprLength b

reverseRule :: Rule -> Rule
reverseRule (Rule (a, b)) = Rule (b, a)

toExpr :: Rule -> Expr
toExpr (Rule (a,b)) = Expr ("=", a, b)

toRule :: Expr -> Maybe Rule
toRule (Expr ("=", a, b)) = Just (Rule (a, b))
toRule _ = Nothing
{-

The rewriting is done by attempting to bind the pattern to a given expression,
and on success, evaluating the substitution in that binding.

If the pattern doesn't match the expression, rewriting will return the original
expression.

E.g. applying the above rule "a o b = a + b + 1" to "((a + b) o b)"
    applyRule rule (Expr ("o", Expr ("+", A, B), B))
would result in
    Expr ("+", Expr ("+", Expr ("+", A, B), B), Literal 1)

-}
applyRule :: Rule -> Expr -> Expr
applyRule (Rule (pat, sub)) expression =
    maybe expression id (evalExpr binding sub)
    where binding = bindPattern pat expression

applyEquality :: Rule -> Expr -> Expr
applyEquality r@(Rule (pat, sub)) expr =
    if matchPattern pat expr
        then applyRule r expr
        else applyRule (reverseRule r) expr

data Binding = Binding (Maybe Expr, Maybe Expr, Maybe Expr) | Invalid

emptyBinding :: Binding
emptyBinding = Binding (Nothing, Nothing, Nothing)

invalidBinding :: Binding -> Bool
invalidBinding Invalid = True
invalidBinding _ = False

evalExpr :: Binding -> Expr -> Maybe Expr
evalExpr binding e | invalidBinding binding = Nothing
evalExpr binding e = Just (mapExpr (applyBinding binding) e)

mapExpr :: (Expr -> Expr) -> Expr -> Expr
mapExpr f (Expr (o, l, r)) = Expr (o, mapExpr f l, mapExpr f r)
mapExpr f (Inv (o, e)) = Inv (o, mapExpr f e)
mapExpr f x = f x

exprLength :: Expr -> Int
exprLength (Expr (o, l, r)) = 1 + exprLength l + exprLength r
exprLength (Inv (o, e)) = 1 + exprLength e
exprLength x = 1

applyBinding :: Binding -> Expr -> Expr
applyBinding (Binding (Just a,_,_)) A = a
applyBinding (Binding (_,Just b,_)) B = b
applyBinding (Binding (_,_,Just c)) C = c
applyBinding _ e = e

bindPattern :: Pattern -> Expr -> Binding
bindPattern pat expr =
    updateBinding emptyBinding pat expr

updateBinding :: Binding -> Pattern -> Expr -> Binding
updateBinding Invalid _ _ = Invalid

updateBinding (Binding (Nothing, b, c)) A a = Binding (Just a, b, c)
updateBinding (Binding (Just a, b, c)) A ae | a == ae = Binding (Just a, b, c)
updateBinding _ A _ = Invalid

updateBinding (Binding (a, Nothing, c)) B b = Binding (a, Just b, c)
updateBinding (Binding (a, Just b, c)) B be | b == be = Binding (a, Just b, c)
updateBinding _ B _ = Invalid

updateBinding (Binding (a, b, Nothing)) C c = Binding (a, b, Just c)
updateBinding (Binding (a, b, Just c)) C ce | c == ce = Binding (a, b, Just c)
updateBinding _ C _ = Invalid

updateBinding b (Literal x) (Literal y) | x == y = b
updateBinding _ (Literal _) _ = Invalid

updateBinding b (Inv (op, p)) (Inv (ope, e)) | op == ope = updateBinding b p e
updateBinding _ (Inv _) _ = Invalid

updateBinding b (Expr (op, pl, pr)) (Expr (ope, el, er)) | op == ope =
    updateBinding (updateBinding b pl el) pr er
updateBinding _ (Expr _) _ = Invalid

updateBinding b (Neutral op) (Neutral ope) | op == ope = b
updateBinding b (Neutral _) _ = Invalid

matchPattern :: Pattern -> Expr -> Bool
matchPattern pat expr = not $ invalidBinding (bindPattern pat expr)

a `plus` b = Expr ("+", a, b)
a `o` b = Expr ("o", a, b)
a `eq` b = Rule (a, b)

{-
{Indexing}

To apply a rewrite rule to a particular place in an equation, you need to be
able to index the symbols in the equation from left to right.

E.g.
    applyRuleAt 1 ((A `plus` B) `eq` (A `o` B)) ((A `plus` Literal 1) `plus` B)
would result in
    ((a o 1) + b) instead of ((a + 1) o b)
-}

applyRuleAt :: Int -> Rule -> Expr -> Expr
applyRuleAt idx rule expr =
    outerMapExprWithIndex (\e i -> if i == idx then applyRule rule e else e) expr

applyEqualityAt :: Int -> Rule -> Expr -> Expr
applyEqualityAt idx rule expr =
    outerMapExprWithIndex (\e i -> if i == idx then applyEquality rule e else e) expr
{-

Doing the indexed rule application with an outer map works, but isn't very
pretty or effective. The outer map can also be put into an infinite loop
(not that applyRuleAt does that.)

An alternative way would be to have a list presentation for the expressions and
work with them. Or mapAtIndex.

-}

subExprAt :: Int -> Expr -> Maybe Expr
subExprAt idx expr = fst $ subExprAt' expr idx 0

subExprAt' :: Expr -> Int -> Int -> (Maybe Expr, Int)
subExprAt' _ i c | c > i = (Nothing, 0)
subExprAt' e@(Expr (o,l,r)) idx count =
    let (mex, c) = subExprAt' l idx count in
    case mex of
        Just ex -> (mex,c)
        Nothing ->
            if c == idx
                then (Just e, c)
                else subExprAt' r idx (c+1)
subExprAt' e i c | i == c = (Just e, c)
subExprAt' (Inv (o,e)) idx count = subExprAt' e idx (count+1)
subExprAt' e i c = (Nothing, c+1)

outerMapExprWithIndex :: (Expr -> Int -> Expr) -> Expr -> Expr
outerMapExprWithIndex f expr = snd $ outerMapExprWithIndex' f expr 0

outerMapExprWithIndex' :: (Expr -> Int -> Expr) -> Expr -> Int -> (Int, Expr)
outerMapExprWithIndex' f (Expr (o,a,b)) c =
    let (c', left) = outerMapExprWithIndex' f a c in
    let (c'', expr) = (c'+1, f (Expr (o, left, b)) c') in
    case expr of
        Expr (o',left',b') -> let (c''', right) = outerMapExprWithIndex' f b' c'' in
                           (c''', Expr (o', left', right))
        x -> outerMapExprWithIndex' f x c''

outerMapExprWithIndex' f (Inv (o,e)) c =
    let (c', e') = (c+1, f e' c) in
    case e' of
        Inv (o',e'') -> let (c'', e''') = outerMapExprWithIndex' f e'' c' in
                        (c'', Inv (o', e'''))
        x -> outerMapExprWithIndex' f x c'

outerMapExprWithIndex' f x c = (c+1, f x c)

File: algebra.js
// Algebra.js - Port of Algebra.hs
// Handles expression manipulation and rules

// Haskell source:
// data Expr = Expr (Op, Expr, Expr)
//           | A | B | C | X | Y | Z
//           | Inv (Op, Expr)
//           | Neutral Op
//           | Literal Int
class Expr {
    constructor(type, ...args) {
        this.type = type;
        this.args = args;
    }

    equals(other) {
        if (!(other instanceof Expr)) return false;
        if (this.type !== other.type) return false;
        if (this.args.length !== other.args.length) return false;
        return this.args.every((arg, i) => {
            if (arg instanceof Expr) return arg.equals(other.args[i]);
            return arg === other.args[i];
        });
    }

    toString() {
        switch(this.type) {
            case 'Op':
                return `(${this.args[1].toString()} ${this.args[0]} ${this.args[2].toString()})`;
            case 'A': return 'a';
            case 'B': return 'b';
            case 'C': return 'c';
            case 'X': return 'x';
            case 'Y': return 'y';
            case 'Z': return 'z';
            case 'Literal': return this.args[0].toString();
            case 'Inv': return `inv(${this.args[0]},${this.args[1].toString()})`;
            case 'Neutral': return `e(${this.args[0]})`;
            default: return '';
        }
    }

    clone() {
        const clonedArgs = this.args.map(arg => 
            arg instanceof Expr ? arg.clone() : arg
        );
        return new Expr(this.type, ...clonedArgs);
    }
}

// Factory functions for expressions
const A = new Expr('A');
const B = new Expr('B');
const C = new Expr('C');
const X = new Expr('X');
const Y = new Expr('Y');
const Z = new Expr('Z');

function Op(op, left, right) {
    return new Expr('Op', op, left, right);
}

function Literal(n) {
    return new Expr('Literal', n);
}

function Inv(op, expr) {
    return new Expr('Inv', op, expr);
}

function Neutral(op) {
    return new Expr('Neutral', op);
}

// Haskell source:
// data Rule = Rule (Pattern, Substitution)
class Rule {
    constructor(left, right) {
        this.left = left;
        this.right = right;
    }

    equals(other) {
        return other instanceof Rule && 
               this.left.equals(other.left) && 
               this.right.equals(other.right);
    }

    toString() {
        return `${this.left.toString()} = ${this.right.toString()}`;
    }

    reverse() {
        return new Rule(this.right, this.left);
    }

    toExpr() {
        return Op("=", this.left, this.right);
    }

    length() {
        return this.left.length() + 1 + this.right.length();
    }
}

// Expression manipulation
Expr.prototype.length = function() {
    if (this.type === 'Op') {
        return 1 + this.args[1].length() + this.args[2].length();
    }
    return 1;
};

// Pattern matching and binding
class Binding {
    constructor() {
        this.bindings = { A: null, B: null, C: null, X: null, Y: null, Z: null };
        this.invalid = false;
    }

    bind(varName, expr) {
        if (this.invalid) return this;
        
        if (this.bindings[varName] === null) {
            this.bindings[varName] = expr;
        } else if (!this.bindings[varName].equals(expr)) {
            this.invalid = true;
        }
        return this;
    }

    apply(expr) {
        if (this.invalid) return null;
        return this.applyToExpr(expr);
    }

    applyToExpr(expr) {
        if (expr.type in this.bindings && this.bindings[expr.type] !== null) {
            return this.bindings[expr.type].clone();
        }
        if (expr.type === 'Op') {
            return Op(expr.args[0], 
                     this.applyToExpr(expr.args[1]), 
                     this.applyToExpr(expr.args[2]));
        }
        if (expr.type === 'Inv') {
            return Inv(expr.args[0], this.applyToExpr(expr.args[1]));
        }
        return expr.clone();
    }
}

// Haskell source:
// matchPattern :: Pattern -> Expr -> Bool
// matchPattern pat expr = not $ invalidBinding (bindPattern pat expr)
function matchPattern(pattern, expr) {
    const binding = new Binding();
    const matches = matchPatternHelper(pattern, expr, binding);
    return matches && !binding.invalid;
}

function matchPatternHelper(pattern, expr, binding) {
    if (binding.invalid) return false;

    if (['A', 'B', 'C', 'X', 'Y', 'Z'].includes(pattern.type)) {
        binding.bind(pattern.type, expr);
        return !binding.invalid;
    }

    if (pattern.type === 'Literal') {
        return expr.type === 'Literal' && pattern.args[0] === expr.args[0];
    }

    if (pattern.type === 'Neutral') {
        return expr.type === 'Neutral' && pattern.args[0] === expr.args[0];
    }

    if (pattern.type === 'Inv') {
        return expr.type === 'Inv' && 
               pattern.args[0] === expr.args[0] &&
               matchPatternHelper(pattern.args[1], expr.args[1], binding);
    }

    if (pattern.type === 'Op') {
        return expr.type === 'Op' && 
               pattern.args[0] === expr.args[0] &&
               matchPatternHelper(pattern.args[1], expr.args[1], binding) &&
               matchPatternHelper(pattern.args[2], expr.args[2], binding);
    }

    return false;
}

// Internal function that returns binding for applyRule
function bindPattern(pattern, expr) {
    const binding = new Binding();
    matchPatternHelper(pattern, expr, binding);
    return binding;
}

// Haskell source:
// applyRule :: Rule -> Expr -> Expr
// applyRule (Rule (pat, sub)) expression =
//     maybe expression id (evalExpr binding sub)
//     where binding = bindPattern pat expression
function applyRule(rule, expr) {
    const binding = bindPattern(rule.left, expr);
    if (!binding.invalid) {
        const result = binding.apply(rule.right);
        if (result !== null) {
            return result;
        }
    }
    return expr.clone();
}

// Haskell source:
// applyEquality :: Rule -> Expr -> Expr
// applyEquality r@(Rule (pat, _)) expr =
//     if matchPattern pat expr
//         then applyRule r expr
//         else applyRule (reverseRule r) expr
function applyEquality(rule, expr) {
    if (matchPattern(rule.left, expr)) {
        return applyRule(rule, expr);
    } else {
        return applyRule(rule.reverse(), expr);
    }
}

// Haskell source:
// applyEqualityAt :: Int -> Rule -> Expr -> Expr
// applyEqualityAt idx rule expr =
//     outerMapExprWithIndex (\e i -> if i == idx then applyEquality rule e else e) expr
function applyEqualityAt(idx, rule, expr) {
    let currentIdx = 0;
    
    function applyAtIndex(e) {
        if (e.type === 'Op') {
            const leftResult = applyAtIndex(e.args[1]);
            const leftIdx = currentIdx;
            
            if (currentIdx === idx) {
                currentIdx++;
                return applyEquality(rule, e);
            }
            currentIdx++;
            
            const rightResult = applyAtIndex(e.args[2]);
            return Op(e.args[0], leftResult, rightResult);
        } else {
            if (currentIdx === idx) {
                currentIdx++;
                return applyEquality(rule, e);
            }
            currentIdx++;
            return e.clone();
        }
    }
    
    return applyAtIndex(expr);
}

// Haskell source:
// subExprAt :: Int -> Expr -> Maybe Expr
function subExprAt(idx, expr) {
    let currentIdx = 0;
    
    function findAtIndex(e) {
        if (e.type === 'Op') {
            const leftResult = findAtIndex(e.args[1]);
            if (leftResult) return leftResult;
            
            if (currentIdx === idx) {
                currentIdx++;
                return e;
            }
            currentIdx++;
            
            return findAtIndex(e.args[2]);
        } else {
            if (currentIdx === idx) {
                currentIdx++;
                return e;
            }
            currentIdx++;
            return null;
        }
    }
    
    return findAtIndex(expr);
}

// Haskell source:
// isTrue :: Rule -> Bool
// isTrue (Rule (a,b)) = a == b
function isTrue(rule) {
    return rule.left.equals(rule.right);
}

// Haskell source:
// isBinding :: Expr -> Rule -> Bool
// isBinding e (Rule (a,b)) = a == b || e == a || e == b
function isBinding(expr, rule) {
    return isTrue(rule) || rule.left.equals(expr) || rule.right.equals(expr);
}

// Haskell source:
// replaceABCwithXYZ :: Rule -> Rule
// replaceABCwithXYZ (Rule (l,r)) = Rule (mapExpr exprABCtoXYZ l, mapExpr exprABCtoXYZ r)
// exprABCtoXYZ A = X
// exprABCtoXYZ B = Y
// exprABCtoXYZ C = Z
function replaceABCwithXYZ(rule) {
    function replace(expr) {
        if (expr.type === 'A') return X;
        if (expr.type === 'B') return Y;
        if (expr.type === 'C') return Z;
        if (expr.type === 'Op') {
            return Op(expr.args[0], replace(expr.args[1]), replace(expr.args[2]));
        }
        if (expr.type === 'Inv') {
            return Inv(expr.args[0], replace(expr.args[1]));
        }
        return expr.clone();
    }
    
    return new Rule(replace(rule.left), replace(rule.right));
}

// Haskell source:
// associativity :: Op -> CheckableRule
// associativity op = (isTrueC, (A `ox` (B `ox` C)) `eq` ((A `ox` B) `ox` C))
function associativity(op) {
    const left = Op(op, A, Op(op, B, C));
    const right = Op(op, Op(op, A, B), C);
    return { predicate: 'bothEqual', rule: new Rule(left, right) };
}

// Haskell source:
// rightNeutral :: Op -> CheckableRule
// rightNeutral op = (isNeutralC (Neutral op), (A `ox` Neutral op) `eq` A)
function rightNeutral(op) {
    const left = Op(op, A, Neutral(op));
    const right = A;
    return { predicate: 'bindNeutral', rule: new Rule(left, right), expr: Neutral(op) };
}

// Haskell source:
// leftNeutral :: Op -> CheckableRule
// leftNeutral op = (isNeutralC (Neutral op), (Neutral op `ox` A) `eq` A)
function leftNeutral(op) {
    const left = Op(op, Neutral(op), A);
    const right = A;
    return { predicate: 'bindNeutral', rule: new Rule(left, right), expr: Neutral(op) };
}

// Haskell source:
// rightInverse :: Op -> CheckableRule
// rightInverse op = (isInverseC (Inv (op, A)), (A `ox` Inv (op, A)) `eq` Neutral op)
function rightInverse(op) {
    const left = Op(op, A, Inv(op, A));
    const right = Neutral(op);
    return { predicate: 'bindInverse', rule: new Rule(left, right), expr: Inv(op, A) };
}

// Haskell source:
// leftInverse :: Op -> CheckableRule
// leftInverse op = (isInverseC (Inv (op, A)), (Inv (op, A) `ox` A) `eq` Neutral op)
function leftInverse(op) {
    const left = Op(op, Inv(op, A), A);
    const right = Neutral(op);
    return { predicate: 'bindInverse', rule: new Rule(left, right), expr: Inv(op, A) };
}

// Haskell source:
// commutativity :: Op -> CheckableRule
// commutativity op = (isTrueC, (A `ox` B) `eq` (B `ox` A))
function commutativity(op) {
    const left = Op(op, A, B);
    const right = Op(op, B, A);
    return { predicate: 'bothEqual', rule: new Rule(left, right) };
}

// Haskell source:
// abelianGroup :: Op -> [CheckableRule]
// abelianGroup op = group op ++ [commutativity op]
// group op = monoid op ++ [rightInverse op, leftInverse op]
// monoid op = semiGroup op ++ [rightNeutral op, leftNeutral op]
// semiGroup op = magma op ++ [associativity op]
function abelianGroup(op) {
    return [
        associativity(op),
        rightNeutral(op),
        leftNeutral(op),
        rightInverse(op),
        leftInverse(op),
        commutativity(op)
    ];
}

// Equality transforms
function eqTrans(op, c) {
    const left = Op("=", Op(op, X, c), Op(op, Y, c));
    const right = Op("=", X, Y);
    return new Rule(left, right);
}

function getUniqueOps(expr) {
    const ops = new Set();
    
    function collect(e) {
        if (e.type === 'Op') {
            ops.add(e.args[0]);
            collect(e.args[1]);
            collect(e.args[2]);
        }
    }
    
    collect(expr);
    return Array.from(ops);
}

function getUniqueVariables(expr) {
    const vars = new Set();
    
    function collect(e) {
        if (['A', 'B', 'C', 'X', 'Y', 'Z'].includes(e.type)) {
            vars.add(e);
        } else if (e.type === 'Op') {
            collect(e.args[1]);
            collect(e.args[2]);
        } else if (e.type === 'Inv') {
            collect(e.args[1]);
        }
    }
    
    collect(expr);
    return Array.from(vars);
}

function equalityTransforms(expr) {
    if (expr.type !== 'Op' || expr.args[0] !== '=') return [];
    
    const transforms = [];
    const ops = getUniqueOps(expr).filter(op => op !== '=');
    const vars = getUniqueVariables(expr);
    
    for (const op of ops) {
        for (const v of vars) {
            transforms.push(eqTrans(op, v));
            if (v.type !== 'Inv') {
                transforms.push(eqTrans(op, Inv(op, v)));
            }
        }
    }
    
    return transforms;
}

// Find matching equalities
function findMatchingEqualities(expr, inventory) {
    return inventory.filter(rule => {
        const binding1 = matchPattern(rule.left, expr);
        const binding2 = matchPattern(rule.right, expr);
        return (binding1 && !binding1.invalid) || (binding2 && !binding2.invalid);
    });
}

function inventoryFor(idx, checkableRule, inventory) {
    const ruleExpr = checkableRule.rule.toExpr();
    const subExpr = subExprAt(idx, ruleExpr);
    
    if (!subExpr) return [];
    
    if (subExpr.type === 'Op' && subExpr.args[0] === '=') {
        return equalityTransforms(subExpr);
    }
    
    return findMatchingEqualities(subExpr, inventory);
}
